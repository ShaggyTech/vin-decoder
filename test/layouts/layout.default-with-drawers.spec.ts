import { mount, createLocalVue } from '@vue/test-utils';
import Vuetify from 'vuetify';
import defaultLayoutWithDrawers from '@/layouts/default-with-drawers.vue';
import vuetifyConfig from '@/plugins/vuetify';

const localVue = createLocalVue();

jest.mock('@/compositions/useLocalStorage', () => ({
  syncHistoryOnMounted: () => jest.fn()
}));

jest.mock('@/compositions/useRootStore', () =>
  jest.requireActual('@/compositions/useRootStore')
);

const mockDrawerItems = [
  {
    icon: 'test-icon-1',
    title: 'Test Page 1',
    to: '/'
  },
  {
    icon: 'test-icon-2',
    title: 'Test Page 2',
    to: '/testpage2'
  }
];

const getMocks = () => {
  return {
    $accessor: {
      _drawer: false,
      get drawer(): boolean {
        return this._drawer;
      },
      set drawer(newDrawer: boolean) {
        this._drawer = newDrawer;
      },

      _rightDrawer: false,
      get rightDrawer(): boolean {
        return this._rightDrawer;
      },
      set rightDrawer(newDrawer: boolean) {
        this._rightDrawer = newDrawer;
      },

      drawerItems: [...mockDrawerItems],

      setDrawer(newDrawer: boolean) {
        return (this._drawer = newDrawer);
      },
      setRightDrawer(newDrawer: boolean) {
        return (this._rightDrawer = newDrawer);
      }
    }
  };
};

const factory = (options: { [propName: string]: any } = {}) => {
  return mount(defaultLayoutWithDrawers, {
    stubs: ['nuxt', 'router-link'],
    localVue,
    ...options
  });
};

describe('Layouts - default', () => {
  let vuetify;
  let wrapper: ReturnType<typeof factory>;

  beforeEach(() => {
    vuetify = new Vuetify(vuetifyConfig);
    wrapper = factory({
      mocks: { ...getMocks() },
      vuetify
    });
  });

  test('component matches snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  test('component has "right" ref', () => {
    expect(wrapper.vm.$data.right).toBeDefined();
  });

  test('component has "title" ref', () => {
    expect(wrapper.vm.$data.title).toBeDefined();
  });

  test('component has "drawer" computed property', () => {
    expect(wrapper.vm.$data.drawer).toBeDefined();
    expect(wrapper.vm.$data.drawer).toEqual(false);
  });

  test('component has "drawerItems" computed property', () => {
    expect(wrapper.vm.$data.drawerItems).toBeDefined();
    expect(wrapper.vm.$data.drawerItems).toEqual(mockDrawerItems);
  });

  test('component has "rightDrawer" computed property', () => {
    expect(wrapper.vm.$data.rightDrawer).toBeDefined();
    expect(wrapper.vm.$data.rightDrawer).toEqual(false);
  });

  test('component has "toggleDrawer" method', () => {
    expect(wrapper.vm.$data.toggleDrawer).toBeDefined();
  });

  test('component has "toggleRightDrawer" method', () => {
    expect(wrapper.vm.$data.toggleRightDrawer).toBeDefined();
  });

  test('"drawer" computed value can be updated via "toggleDrawer" method', () => {
    const toggleDrawer: Function = wrapper.vm.$data.toggleDrawer;
    const getDrawer = (): boolean => wrapper.vm.$data.drawer;

    expect(getDrawer()).toEqual(false);

    toggleDrawer(getDrawer());

    expect(getDrawer()).toEqual(true);
  });

  test('"rightDrawer" computed value can be updated via "toggleRightDrawer" method', () => {
    const toggleRightDrawer: Function = wrapper.vm.$data.toggleRightDrawer;
    const getRightDrawer = (): boolean => wrapper.vm.$data.rightDrawer;

    expect(getRightDrawer()).toEqual(false);

    toggleRightDrawer(getRightDrawer());

    expect(getRightDrawer()).toEqual(true);
  });
});
