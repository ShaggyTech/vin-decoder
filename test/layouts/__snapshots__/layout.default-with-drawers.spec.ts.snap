// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layouts - default component matches snapshot 1`] = `
<div
  class="v-application v-application--is-ltr theme--dark"
  data-app="true"
  id="app"
>
  <div
    class="v-application--wrap"
  >
    <nav
      class="v-navigation-drawer v-navigation-drawer--close v-navigation-drawer--fixed v-navigation-drawer--is-mobile theme--dark"
      style="height: 100vh; top: 0px; transform: translateX(-100%); width: 256px;"
    >
      <div
        class="v-navigation-drawer__content"
      >
        <div
          class="v-list v-sheet theme--dark"
        >
          <router-link-stub
            class="v-list-item v-list-item--link theme--dark"
            router=""
            tabindex="0"
          >
            <div
              class="v-list-item__action"
            >
              <i
                aria-hidden="true"
                class="v-icon notranslate test test-icon-1 theme--dark"
              />
            </div>
             
            <div
              class="v-list-item__content"
            >
              <div
                class="v-list-item__title"
              >
                Test Page 1
              </div>
            </div>
          </router-link-stub>
          <router-link-stub
            class="v-list-item v-list-item--link theme--dark"
            router=""
            tabindex="0"
          >
            <div
              class="v-list-item__action"
            >
              <i
                aria-hidden="true"
                class="v-icon notranslate test test-icon-2 theme--dark"
              />
            </div>
             
            <div
              class="v-list-item__content"
            >
              <div
                class="v-list-item__title"
              >
                Test Page 2
              </div>
            </div>
          </router-link-stub>
        </div>
      </div>
      <div
        class="v-navigation-drawer__border"
      />
    </nav>
     
    <header
      class="v-sheet theme--dark v-toolbar v-app-bar v-app-bar--fixed"
      style="height: 64px; margin-top: 0px; transform: translateY(0px); left: 0px; right: 0px;"
    >
      <div
        class="v-toolbar__content"
        style="height: 64px;"
      >
        <button
          class="v-app-bar__nav-icon v-btn v-btn--flat v-btn--icon v-btn--round theme--dark v-size--default"
          type="button"
        >
          <span
            class="v-btn__content"
          >
            <i
              aria-hidden="true"
              class="v-icon notranslate mdi mdi-menu theme--dark"
            />
          </span>
        </button>
         
        <div
          class="spacer"
        />
         
        <div
          class="v-toolbar__title text-center"
        >
          ShaggyTech.com - VIN Decoder
        </div>
         
        <div
          class="spacer"
        />
         
        <button
          class="v-btn v-btn--flat v-btn--icon v-btn--round theme--dark v-size--default"
          type="button"
        >
          <span
            class="v-btn__content"
          >
            <i
              aria-hidden="true"
              class="v-icon notranslate mdi mdi-menu theme--dark"
            />
          </span>
        </button>
      </div>
    </header>
     
    <main
      class="v-main grey lighten-1"
      style="padding: 64px 0px 0px 0px;"
    >
      <div
        class="v-main__wrap"
      >
        <nuxt-stub />
      </div>
    </main>
     
    <aside
      class="v-navigation-drawer v-navigation-drawer--close v-navigation-drawer--fixed v-navigation-drawer--is-mobile v-navigation-drawer--right v-navigation-drawer--temporary theme--dark"
      style="height: 100%; top: 0px; transform: translateX(100%); width: 256px;"
    >
      <div
        class="v-navigation-drawer__content"
      >
        <div
          class="v-list v-sheet theme--dark"
          role="list"
        >
          <div
            class="v-list-item theme--dark"
            role="listitem"
            tabindex="-1"
          >
            <div
              class="v-list-item__action"
            >
              <i
                aria-hidden="true"
                class="v-icon notranslate mdi mdi-repeat theme--light"
              />
            </div>
             
            <div
              class="v-list-item__title"
            >
              Switch drawer (click me)
            </div>
          </div>
        </div>
      </div>
      <div
        class="v-navigation-drawer__border"
      />
    </aside>
     
    <footer
      class="v-footer text-center v-sheet theme--dark v-footer--fixed"
      style="left: 0px; right: 0px; bottom: 0px;"
    >
      <div
        class="text-center col col-12"
      >
        
      Made with ❤ by Brandon Eichler - ©
      2020
    
      </div>
    </footer>
  </div>
</div>
`;
